### PR TITLE
Make state changes in reducer more comprehensible.

### PR DESCRIPTION
**Pull request by nothingnesses**
I tried to make the states more explicit by using a `State` enum which should make it easier to keep track of their transitions.

### CODE REVIEW

File: src/compiler/compile.rs [hunk 1]
```diff
@@ -65,46 +65,44 @@ pub fn build_rulebook(book: &language::rulebook::RuleBook) -> (String, String) {
...
+  line(&mut fast_visit, 6, &format!("}}"));
 
   // fast apply
   let mut fast_apply = String::new();
-  line(&mut fast_apply, 8, &format!("match fid {{"));
+  line(&mut fast_apply, 6, &format!("match fid {{"));
   for id in itertools::sorted(book.id_to_arit.keys()) {
     if id >= &runtime::PRECOMP_COUNT {
       let name = book.id_to_name.get(id).unwrap();
       let rules =
```
Comment (nothingnesses) [inline src/compiler/compile.rs:91]: I think this line seems incorrect since it currently just sets `rules` to `()`.

**Comment by VictorTaelin** [comment]
Are you absolutely sure that is compiled to the same assembly? The reason the loops are so contrived is that Rust doesn't have GOTO, making it impossible to make direct jumps. Sadly, pattern-matching in a state variable was NOT being optimized by Rust's compiler to direct jumps (at least I couldn't get it to). This impacted the performance by a few %s - nothing huge, but still impactful. I spent a long time (as in, days) trying to get it compile to ASM that jumps directly, and by using these nested loops I managed to avoid the pattern-matching, at the cost of some (admittedly) ugly / hard to follow code.

**Comment by nothingnesses** [comment]
I ran benchmarks and you're correct, my version was unfortunately slower:

<details>
<summary>Using nested loops:</summary>

```
   Compiling main v1.0.0-beta (/home/.shared/repositories/github/nothingnesses/LambdaVM-Rust/bench/work/main)
    Finished release [optimized] target(s) in 1m 39s
HVM-1 | sort/bitonic | 1 | 0.233s
HVM-1 | sort/bitonic | 2 | 0.253s
HVM-1 | sort/bitonic | 3 | 0.251s
HVM-1 | sort/bitonic | 4 | 0.251s
HVM-1 | sort/bitonic | 5 | 0.243s
HVM-1 | sort/bitonic | 6 | 0.245s
HVM-1 | sort/bitonic | 7 | 0.243s
HVM-1 | sort/bitonic | 8 | 0.245s
HVM-1 | sort/bitonic | 9 | 0.255s
HVM-1 | sort/bitonic | 10 | 0.271s
HVM-1 | sort/bitonic | 11 | 0.315s
HVM-1 | sort/bitonic | 12 | 0.408s
HVM-1 | sort/bitonic | 13 | 0.629s
HVM-1 | sort/bitonic | 14 | 1.139s
HVM-1 | sort/bitonic | 15 | 2.327s
HVM-1 | sort/bitonic | 16 | 4.833s
HVM-1 | sort/bitonic | 17 | 9.129s
HVM-1 | sort/bitonic | 18 | 17.135s
HVM-1 | sort/bitonic | 19 | 35.380s
HVM-1 | sort/bitonic | 20 | 73.806s
HVM-1 | sort/bitonic | 21 | 161.132s
   Compiling main v1.0.0-beta (/home/.shared/repositories/github/nothingnesses/LambdaVM-Rust/bench/work/main)
    Finished release [optimized] target(s) in 34.05s
HVM-1 | sort/bubble | 12 | 1.770s
HVM-1 | sort/bubble | 13 | 6.299s
HVM-1 | sort/bubble | 14 | 17.508s
HVM-1 | sort/bubble | 15 | 59.821s
   Compiling main v1.0.0-beta (/home/.shared/repositories/github/nothingnesses/LambdaVM-Rust/bench/work/main)
    Finished release [optimized] target(s) in 18.70s
HVM-1 | sort/quick | 1 | 0.243s
HVM-1 | sort/quick | 2 | 0.253s
HVM-1 | sort/quick | 3 | 0.253s
HVM-1 | sort/quick | 4 | 0.242s
HVM-1 | sort/quick | 5 | 0.242s
HVM-1 | sort/quick | 6 | 0.243s
HVM-1 | sort/quick | 7 | 0.245s
HVM-1 | sort/quick | 8 | 0.245s
HVM-1 | sort/quick | 9 | 0.247s
HVM-1 | sort/quick | 10 | 0.256s
HVM-1 | sort/quick | 11 | 0.271s
HVM-1 | sort/quick | 12 | 0.407s
HVM-1 | sort/quick | 13 | 0.397s
HVM-1 | sort/quick | 14 | 0.543s
HVM-1 | sort/quick | 15 | 0.881s
HVM-1 | sort/quick | 16 | 1.624s
HVM-1 | sort/quick | 17 | 3.079s
HVM-1 | sort/quick | 18 | 6.136s
HVM-1 | sort/quick | 19 | 11.005s
HVM-1 | sort/quick | 20 | 20.922s
HVM-1 | sort/quick | 21 | 39.926s
   Compiling main v1.0.0-beta (/home/.shared/repositories/github/nothingnesses/LambdaVM-Rust/bench/work/main)
    Finished release [optimized] target(s) in 19.87s
HVM-1 | sort/radix | 1 | 0.230s
HVM-1 | sort/radix | 2 | 0.242s
HVM-1 | sort/radix | 3 | 0.236s
HVM-1 | sort/radix | 4 | 0.230s
HVM-1 | sort/radix | 5 | 0.252s
HVM-1 | sort/radix | 6 | 0.249s
HVM-1 | sort/radix | 7 | 0.251s
HVM-1 | sort/radix | 8 | 0.257s
HVM-1 | sort/radix | 9 | 0.250s
HVM-1 | sort/radix | 10 | 0.284s
HVM-1 | sort/radix | 11 | 0.326s
HVM-1 | sort/radix | 12 | 0.383s
HVM-1 | sort/radix | 13 | 0.428s
HVM-1 | sort/radix | 14 | 0.615s
HVM-1 | sort/radix | 15 | 0.991s
HVM-1 | sort/radix | 16 | 1.732s
HVM-1 | sort/radix | 17 | 3.209s
HVM-1 | sort/radix | 18 | 5.994s
HVM-1 | sort/radix | 19 | 11.176s
HVM-1 | sort/radix | 20 | 21.209s
HVM-1 | sort/radix | 21 | 52.724s
```
</details>

<details>
<summary>Using `State` enum:</summary>

```
   Compiling main v1.0.0-beta (/home/.shared/repositories/github/nothingnesses/HVM.reducer/bench/work/main)
    Finished release [optimized] target(s) in 20.30s
HVM-1 | sort/bitonic | 1 | 0.240s
HVM-1 | sort/bitonic | 2 | 0.242s
HVM-1 | sort/bitonic | 3 | 0.241s
HVM-1 | sort/bitonic | 4 | 0.240s
HVM-1 | sort/bitonic | 5 | 0.243s
HVM-1 | sort/bitonic | 6 | 0.241s
HVM-1 | sort/bitonic | 7 | 0.245s
HVM-1 | sort/bitonic | 8 | 0.245s
HVM-1 | sort/bitonic | 9 | 0.256s
HVM-1 | sort/bitonic | 10 | 0.279s
HVM-1 | sort/bitonic | 11 | 0.332s
HVM-1 | sort/bitonic | 12 | 0.456s
HVM-1 | sort/bitonic | 13 | 0.763s
HVM-1 | sort/bitonic | 14 | 1.424s
HVM-1 | sort/bitonic | 15 | 2.809s
HVM-1 | sort/bitonic | 16 | 6.081s
HVM-1 | sort/bitonic | 17 | 12.013s
HVM-1 | sort/bitonic | 18 | 23.409s
HVM-1 | sort/bitonic | 19 | 47.397s
HVM-1 | sort/bitonic | 20 | 100.276s
HVM-1 | sort/bitonic | 21 | 219.774s
   Compiling main v1.0.0-beta (/home/.shared/repositories/github/nothingnesses/HVM.reducer/bench/work/main)
    Finished release [optimized] target(s) in 19.50s
HVM-1 | sort/bubble | 12 | 2.222s
HVM-1 | sort/bubble | 13 | 8.023s
HVM-1 | sort/bubble | 14 | 23.961s
HVM-1 | sort/bubble | 15 | 84.230s
   Compiling main v1.0.0-beta (/home/.shared/repositories/github/nothingnesses/HVM.reducer/bench/work/main)
    Finished release [optimized] target(s) in 18.95s
HVM-1 | sort/quick | 1 | 0.243s
HVM-1 | sort/quick | 2 | 0.232s
HVM-1 | sort/quick | 3 | 0.248s
HVM-1 | sort/quick | 4 | 0.250s
HVM-1 | sort/quick | 5 | 0.232s
HVM-1 | sort/quick | 6 | 0.231s
HVM-1 | sort/quick | 7 | 0.232s
HVM-1 | sort/quick | 8 | 0.234s
HVM-1 | sort/quick | 9 | 0.261s
HVM-1 | sort/quick | 10 | 0.276s
HVM-1 | sort/quick | 11 | 0.305s
HVM-1 | sort/quick | 12 | 0.333s
HVM-1 | sort/quick | 13 | 0.510s
HVM-1 | sort/quick | 14 | 0.595s
HVM-1 | sort/quick | 15 | 0.980s
HVM-1 | sort/quick | 16 | 1.646s
HVM-1 | sort/quick | 17 | 3.261s
HVM-1 | sort/quick | 18 | 6.943s
HVM-1 | sort/quick | 19 | 11.862s
HVM-1 | sort/quick | 20 | 21.615s
HVM-1 | sort/quick | 21 | 42.279s
   Compiling main v1.0.0-beta (/home/.shared/repositories/github/nothingnesses/HVM.reducer/bench/work/main)
    Finished release [optimized] target(s) in 19.68s
HVM-1 | sort/radix | 1 | 0.248s
HVM-1 | sort/radix | 2 | 0.249s
HVM-1 | sort/radix | 3 | 0.231s
HVM-1 | sort/radix | 4 | 0.232s
HVM-1 | sort/radix | 5 | 0.249s
HVM-1 | sort/radix | 6 | 0.235s
HVM-1 | sort/radix | 7 | 0.237s
HVM-1 | sort/radix | 8 | 0.243s
HVM-1 | sort/radix | 9 | 0.259s
HVM-1 | sort/radix | 10 | 0.285s
HVM-1 | sort/radix | 11 | 0.330s
HVM-1 | sort/radix | 12 | 0.416s
HVM-1 | sort/radix | 13 | 0.552s
HVM-1 | sort/radix | 14 | 0.817s
HVM-1 | sort/radix | 15 | 1.350s
HVM-1 | sort/radix | 16 | 2.436s
HVM-1 | sort/radix | 17 | 4.501s
HVM-1 | sort/radix | 18 | 8.958s
HVM-1 | sort/radix | 19 | 16.492s
HVM-1 | sort/radix | 20 | 31.762s
HVM-1 | sort/radix | 21 | 75.410s
```
</details>

I also tried to inspect the emitted LLVM-IR to see why that was but I soon gave up as the output was very big and I'm not proficient enough at analysing it to come to a conclusion.

Best keep using nested loops for now.
